pkg/io/jsonio/unmarshal.go

```diff
@@ -28,169 +25,74 @@ func NewReader() *Reader {
-		containedSubnets := make([]ir.ID, 0)
-		overlappingVPCs := make([]ir.ID, 0)
-
-		for _, cidr := range segment {
-			c, err := netset.IPBlockFromCidr(cidr)
-			if err != nil {
-				return nil, err
-			}
-			vpcs := parseOverlappingVpcs(c, configDefs.VPCs)
-			subnets, err := configDefs.SubnetsContainedInCidr(*c)
-			if err != nil {
-				return nil, err
-			}
-
-			cidrs = cidrs.Union(c)
-			overlappingVPCs = append(overlappingVPCs, vpcs...)
-			containedSubnets = append(containedSubnets, subnets...)
-		}
-		cidrSegmentDetails := ir.CidrSegmentDetails{
-			Cidrs:            cidrs,
-			ContainedSubnets: slices.Compact(slices.Sorted(slices.Values(containedSubnets))),
-			OverlappingVPCs:  slices.Compact(slices.Sorted(slices.Values(overlappingVPCs))),
-		}
-		result[segmentName] = &cidrSegmentDetails
-	}
-	return result, nil
-}
+// replace all resources names to fully qualified name
+func replaceResourcesName(jsonSpec *spec.Spec, defs *ir.Definitions) (*spec.Spec, *ir.Definitions, error) {
+	config := defs.ConfigDefs
 
-// read externals from conn-spec
-func translateExternals(m map[string]string) (map[ir.ID]*ir.ExternalDetails, error) {
-	result := make(map[ir.ID]*ir.ExternalDetails)
-	for k, v := range m {
-		address, err := netset.IPBlockFromCidrOrAddress(v)
-		if err != nil {
-			return nil, err
-		}
-		result[k] = &ir.ExternalDetails{ExternalAddrs: address}
-	}
-	return result, nil
-}
+	// calculate distinct and ambiguous names for every endpoint type
+	distinctSubnets, ambiguousSubnets := inverseMapToFullyQualifiedName(config.Subnets)
+	distinctNifs, ambiguousNifs := inverseMapToFullyQualifiedName(config.NIFs)
+	distinctInstances, ambiguousInstances := inverseMapToFullyQualifiedName(config.Instances)
+	distinctVpes, ambiguousVpes := inverseMapToFullyQualifiedName(config.VPEs)
 
-// replace all resources names in conn-spec to fully qualified name
-func replaceResourcesName(jsonSpec *spec.Spec, subnetSegments map[string][]ir.ID,
-	config *ir.ConfigDefs) (*spec.Spec, map[ir.ID]*ir.SubnetSegmentDetails, error) {
-	subnetsCache, ambiguousSubnets := inverseMapToFullyQualifiedName(config.Subnets)
-	nifsCache, ambiguousNifs := inverseMapToFullyQualifiedName(config.NIFs)
-	instancesCache, ambiguousInstances := inverseMapToFullyQualifiedName(config.Instances)
-	vpesCache, ambiguousVpes := inverseMapToFullyQualifiedName(config.VPEs)
+	// translate segments to fully qualified names
+	nifSegments, err2 := replaceSegmentNames(defs.NifSegments, distinctNifs, ambiguousNifs, spec.ResourceType(spec.SegmentTypeNif))
+	vpeSegments, err4 := replaceSegmentNames(defs.VpeSegments, distinctVpes, ambiguousVpes, spec.ResourceType(spec.SegmentTypeVpe))
+	subnetSegments, err1 := replaceSegmentNames(defs.SubnetSegments, distinctSubnets, ambiguousSubnets,
+		spec.ResourceType(spec.SegmentTypeSubnet))
+	instanceSegments, err3 := replaceSegmentNames(defs.InstanceSegments, distinctInstances, ambiguousInstances,
+		spec.ResourceType(spec.SegmentTypeInstance))
```

Comment:
Can we make the index of `err` variables monotonically increasing? (forgive my OCD)